File: src/main/java/ch/leadrian/samp/kamp/textkeygen/TextKeysGeneratorPlugin.java
package ch.leadrian.samp.kamp.textkeygen;

import org.gradle.api.NonNullApi;
import org.gradle.api.Plugin;
import org.gradle.api.Project;

@NonNullApi
public class TextKeysGeneratorPlugin implements Plugin<Project> {

    @Override
    public void apply(Project project) {
        TextKeysGeneratorPluginExtension extension = project
                .getExtensions()
                .create("textKeyGenerator", TextKeysGeneratorPluginExtension.class);
        project.getTasks().create("generateTextKeys", GenerateTextKeysTask.class);
    }
}

File: src/main/kotlin/ch/leadrian/samp/kamp/gradle/plugin/textkeygen/TextKeysGeneratorPlugin.kt
package ch.leadrian.samp.kamp.gradle.plugin.textkeygen

import org.gradle.api.NonNullApi
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.tasks.compile.JavaCompile
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

@NonNullApi
open class TextKeysGeneratorPlugin : Plugin<Project> {

    override fun apply(project: Project) {
        project.extensions.create("textKeyGenerator", TextKeysGeneratorPluginExtension::class.java)
        val generateTextKeysTask = project.tasks.create("generateTextKeys", GenerateTextKeysTask::class.java)
        project.tasks.withType(JavaCompile::class.java) { it.dependsOn(generateTextKeysTask) }
        project.tasks.withType(KotlinCompile::class.java) { it.dependsOn(generateTextKeysTask) }
    }
}
